sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c

```diff
@@ -25,53 +30,174 @@
+ *
+ * @param[in] pkt       A packet with the first snip pointing to the extension
+ *                      header to process.
+ * @param[in] nh        Protocol number of @p pkt.
+ *
+ * @return  The packet for further processing.
+ * @return  NULL, on error or if packet was consumed (by e.g. forwarding via
+ *          a routing header). @p pkt is released in case of error.
+ */
+static gnrc_pktsnip_t *_demux(gnrc_pktsnip_t *pkt, unsigned nh);
+
+gnrc_pktsnip_t *gnrc_ipv6_ext_process_hopopt(gnrc_pktsnip_t *pkt,
+                                             uint8_t *nh)
 {
-    gnrc_pktsnip_t *ipv6;
-    ipv6_ext_t *ext = (ipv6_ext_t *) current->data;
-    size_t current_offset;
-    ipv6_hdr_t *hdr;
-    int res;
+    if ((*nh == PROTNUM_IPV6_EXT_HOPOPT) && (pkt != NULL)) {
+        /* if a hop-by-hop options header is within the IPv6 packet it comes
+         * immediately after the IPv6 header and it must be processed before
+         * the packet is forwarded
+         * (see https://tools.ietf.org/html/rfc8200#section-4.1) */
+        if ((pkt = _demux(pkt, *nh)) == NULL) {
+            DEBUG("ipv6 ext: packet was consumed in extension header "
+                  "handling\n");
+            return NULL;
+        }
+        /* packet can only be unmarked (since we are the first one to
+         * process this packet and 6LoWPAN only provides packets unmarked)
+         *
+         * pkt -----------> pkt->next --------> pkt->next->next
+         *  v                     v                         v
+         * next header          hop-by-hop option       IPv6 header */
+        assert((pkt->next != NULL) && (pkt->next->next != NULL) &&
+               (pkt->next->next->type == GNRC_NETTYPE_IPV6));
+        *nh = ((ipv6_ext_t *)pkt->next->data)->nh;
+        if (_duplicate_hopopt(pkt, *nh)) {
+            return NULL;
+        }
+    }
+    return pkt;
+}
 
-    /* check seg_left early to avoid duplicating the packet */
-    if (((ipv6_ext_rh_t *)ext)->seg_left == 0) {
-        return GNRC_IPV6_EXT_RH_AT_DST;
+gnrc_pktsnip_t *gnrc_ipv6_ext_process_all(gnrc_pktsnip_t *pkt,
+                                          uint8_t *nh)
+{
+    bool is_ext = true;
+    while (is_ext) {
+        switch (*nh) {
+            case PROTNUM_IPV6_EXT_DST:
+            case PROTNUM_IPV6_EXT_RH:
+            case PROTNUM_IPV6_EXT_FRAG:
+            case PROTNUM_IPV6_EXT_AH:
+            case PROTNUM_IPV6_EXT_ESP:
+            case PROTNUM_IPV6_EXT_MOB: {
+                ipv6_ext_t *ext_hdr;
+
+                DEBUG("ipv6: handle extension header (nh = %u)\n", *nh);
+                ext_hdr = pkt->data;
+                if ((pkt = _demux(pkt, *nh)) == NULL) {
+                    DEBUG("ipv6: packet was consumed by extension header "
+                          "handling\n");
+                    return NULL;
+                }
+                *nh = ext_hdr->nh;
+                if (_duplicate_hopopt(pkt, *nh)) {
+                    return NULL;
```

Comment:
The documentation of `@param[in] nh ` of `gnrc_ipv6_ext_process_all()` in `ext.h` does not contain a remark about this case (in contrast to the `gnrc_ipv6_ext_process_hopopt()` doc, which says ` if the return value is NULL the value undefined`). Is this on purpose?

Reply:
Done